File: wrappers/kotlin/lib/src/test/kotlin/org/devolutions/crypto/TestUtils.kt
package org.devolutions.crypto

// Code is ported from this post:
// https://www.geeksforgeeks.org/check-whether-an-array-is-subarray-of-another-array/
fun <T> List<T>.isSubArray(data: List<T>): Boolean {
    var i = 0
    var j = 0

    while (i < data.size && j < this.size) {
        if (data[i] == this[j]) {
            i += 1;
            j += 1;

            if (j == this.size) {
                return true
            }
        } else {
            i = i - j + 1
            j = 0
        }
    }

    return false
}

File: wrappers/dart/test/test_utils.dart
import 'dart:typed_data';

/// Code ported from Kotlin tests
/// https://www.geeksforgeeks.org/check-whether-an-array-is-subarray-of-another-array/
extension ListUtils<T> on List<T> {
  bool isSubArray(List<T> data) {
    int i = 0;
    int j = 0;

    while (i < data.length && j < length) {
      if (data[i] == this[j]) {
        i += 1;
        j += 1;

        if (j == length) {
          return true;
        }
      } else {
        i = i - j + 1;
        j = 0;
      }
    }

    return false;
  }
}

/// Extension for Uint8List comparison
extension Uint8ListUtils on Uint8List {
  bool contentEquals(Uint8List other) {
    if (length != other.length) return false;
    for (int i = 0; i < length; i++) {
      if (this[i] != other[i]) return false;
    }
    return true;
  }
}
